gear-programs/vft-manager/app/src/services/submit_receipt/token_operations.rs

```diff
@@ -43,22 +43,50 @@ where
         .handle_reply(move || handle_reply::<Action>(slot, transaction_index))
         .map_err(|_| Error::ReplyHook)?
         .await
-        .map_err(|_| Error::ReplyFailure)?;
-
-    Ok(())
+        .map_err(|e| match e {
+            GStdError::Timeout(..) => Error::ReplyTimeout,
+            _ => Error::Internal(format!("{e:?}").into_bytes()),
+        })
+        .and_then(|_| {
+            let key = (slot, transaction_index);
+            match super::reply_statuses_mut().remove(&key) {
+                Some(status) => status.clone(),
+                None => {
+                    unreachable!("Status always set if a VFT invocation was successful");
+                }
+            }
+        })
 }
 
 fn handle_reply<Action>(slot: u64, transaction_index: u64)
 where
     Action: ActionIo,
     Action::Reply: Reply,
 {
-    let reply_bytes = msg::load_bytes()
-        .expect("May fail because of the insufficient gas only but the limit was specified by the caller; qed");
-    let reply = Action::decode_reply(&reply_bytes)
-        .expect("May fail only if there is no VFT-program at the specified address; qed");
+    let status = {
+        || {
+            let reply_code = msg::reply_code().map_err(|_| Error::ReplyCode)?;
+            if let ReplyCode::Error(_reason) = reply_code {
+                // An error reply received. The error description will be extracted by the sender.
+                // However, we must set status to `Err` here to skip further processing.
+                return Err(Error::ReplyFailure);
+            }
+            let reply_bytes = msg::load_bytes().map_err(|_| Error::GasForReplyTooLow)?;
```

Comment:
```suggestion
            let reply_bytes = msg::load_bytes().map_err(|e| Error::GasForReplyTooLow(format!("{e:?}"))?;
```

Reply:
again, how necessary is this "embellishment" with a string description given this error can only stem from a single reason?